src/v/storage/disk_log_impl.cc

```diff
@@ -2474,6 +2474,32 @@ auto disk_log_impl::get_file_offset(
     };
 }
 
+model::offset disk_log_impl::cleanly_compacted_prefix_offset() const {
+    _cleanly_compacted_offset = std::max(
+      _cleanly_compacted_offset, _start_offset);
+    if (!_segs.empty()) {
+        auto first_seg_to_check_it = _segs.lower_bound(
+          _cleanly_compacted_offset);
+        auto first_non_clean_seg_it = std::find_if(
+          first_seg_to_check_it, _segs.end(), [](const segment_set::type& seg) {
+              return !seg->index().has_clean_compact_timestamp();
+          });
+        if (first_non_clean_seg_it == _segs.end()) [[unlikely]] {
+            const auto& last_seg = *_segs.back();
+            const auto& last_idx = last_seg.index();
+            // max_offset is known to store rubbish for empty segments
+            _cleanly_compacted_offset = last_seg.empty()
+                                          ? last_idx.base_offset()
+                                          : model::next_offset(
+                                              last_idx.max_offset());
```

Comment:
@WillemKauf @bashtanov looks like this might examine the active segment. are there any concerns with doing that? we have cases in storage where we could look at the active segment, but we wait for it to roll so that its state is more stable.

Reply:
IIRC the last one is never compacted. @WillemKauf does it guard us from touching the newly created empty one?

Reply:
@dotnwat does it look better to you?